lib/srv/usermgmt.go

```diff
@@ -0,0 +1,261 @@
+	CreateGroup(group string) error
+	// CreateUser creates a user on a host.
+	CreateUser(name string, groups []string) error
+	// DeleteUser deletes a user from a host.
+	DeleteUser(name string) error
+}
+
+// HostUsersProvisioningBackend is used to implement HostUsersBackend
+type HostUsersProvisioningBackend struct{}
+
+type userCloser struct {
+	users    HostUsers
+	backend  HostUsersBackend
+	username string
+}
+
+func (u *userCloser) Close() error {
+	teleportGroup, err := u.backend.LookupGroup(types.TeleportServiceGroup)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = u.users.DeleteUser(u.username, teleportGroup.Gid)
+	return trace.Wrap(err)
+}
+
+var ErrUserLoggedIn = errors.New("User logged in error")
+
+type HostUsers interface {
+	// CreateUser creates a temporary Teleport user in the TeleportServiceGroup
+	CreateUser(name string, hostRoleInfo *services.HostUsersInfo) (*user.User, io.Closer, error)
+	// DeleteUser deletes a temporary Teleport user only if they are
+	// in a specified group
+	DeleteUser(name string, gid string) error
+	// DeleteAllUsers deletes all suer in the TeleportServiceGroup
+	DeleteAllUsers() error
+	// UserCleanup starts a periodic user deletion cleanup loop for
+	// users that failed to delete
+	UserCleanup()
+	// Shutdown cancels the UserCleanup loop
+	Shutdown()
+
+	// UserExists returns nil should a hostuser exist
+	UserExists(string) (*user.User, error)
+}
+
+type HostUserManagement struct {
+	backend HostUsersBackend
+	ctx     context.Context
+	cancel  context.CancelFunc
+	storage *local.PresenceService
+	lease   *types.SemaphoreLease
+}
+
+var _ HostUsers = &HostUserManagement{}
+
+// CreateUser creates a temporary Teleport user in the TeleportServiceGroup
+func (u *HostUserManagement) CreateUser(name string, ui *services.HostUsersInfo) (*user.User, io.Closer, error) {
+	tempUser, err := u.backend.Lookup(name)
+	if err != nil && err != user.UnknownUserError(name) {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	if tempUser != nil {
+		// try to delete even if the user already exists as only users
+		// in the teleport-system group will be deleted and this way
+		// if a user creates multiple sessions the account will
+		// succeed in deletion
+		return tempUser, &userCloser{
+			username: name,
+			users:    u,
+			backend:  u.backend,
+		}, trace.AlreadyExists("User %q already exists", name)
+	}
+
+	groups := append(ui.Groups, types.TeleportServiceGroup)
+	var errs []error
+	for _, group := range groups {
+		if err := u.createGroupIfNotExist(group); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+	}
+	if err := trace.NewAggregate(errs...); err != nil {
+		return nil, nil, trace.WrapWithMessage(err, "error while creating groups")
+	}
+
+	if err := u.storage.UpsertUserLoginTime(u.ctx, name, time.Now()); err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	err = u.backend.CreateUser(name, groups)
+	if err != nil && !trace.IsAlreadyExists(err) {
```

Comment:
We should hold a semaphore lock while doing this, otherwise there's still nothing that prevents the cleanup loop from reading an old login time, then here we update the login time, create (an existing) user, and then the cleanup loop goes ahead and deletes the actual host user.

We don't need a lease to run the cleanup loop at all if instead we wait to grab the "host user manipulation lock" while doing either `UpsertUserLoginTime`+`CreateUser` or `GetUserLoginTime`+`DeleteUser`.